firewall: move rpfilter to mangle.PREROUTING to fix conntrack

milahu opened the pull request:
fix #51258 

make `conntrack` work: `rpfilter` must be in the `mangle` table, after the `conntrack` stage

see [the netfilter flowchart](https://upload.wikimedia.org/wikipedia/commons/3/37/Netfilter-packet-flow.svg): input path, network layer

```
--> raw.prerouting --> conntrack --> mangle.prerouting -->
```

###### Motivation for this change

this patch is needed to make VPN work (wireguard protocol, configured with `wg-quick`, but probably also with openvpn protocol)
otherwise the VPN server is not reachable (response packets are dropped)

workarounds:
* run `ip route add $vpn_server_ip via $local_gateway_ip`, for example `ip route add 1.2.3.4 via 192.168.0.1`. this works, because this explcit route has a higher priority in the `wg-quick` routing config, so the return-path is found correctly
* in `configuration.nix` set `networking.firewall.checkReversePath = false;` to disable `rpfilter` completely (security risk!)

not tested yet, `nixos-rebuild switch` fails to compile (for other reasons than this patch)

thanks to `another` from `#wireguard` at `freenode.net` for finding the bug


<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->



###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/is-there-a-way-to-extend-the-default-list-of-modules-imported-via-an-overlay-of-some-sort/5282/6

milahu commented:
<details>
<summary>tested with success : )</summary>

before patch, with `networking.firewall.checkReversePath = true;` (default config)
```
# drop the new rules
iptables -t mangle -D PREROUTING -j nixos-fw-rpfilter
iptables -t mangle -F nixos-fw-rpfilter
iptables -t mangle -X nixos-fw-rpfilter

nixos-rebuild switch

# table mangle
iptables -t mangle -L | grep rpfilter

# table raw
iptables -t raw -L | grep rpfilter
nixos-fw-rpfilter  all  --  anywhere             anywhere            
Chain nixos-fw-rpfilter (1 references)
RETURN     all  --  anywhere             anywhere             rpfilter validmark

```

after patch
```
# drop the old rules
iptables -t raw -D PREROUTING -j nixos-fw-rpfilter
iptables -t raw -F nixos-fw-rpfilter
iptables -t raw -X nixos-fw-rpfilter

nixos-rebuild switch

# table mangle
iptables -t mangle -L | grep rpfilter
nixos-fw-rpfilter  all  --  anywhere             anywhere            
Chain nixos-fw-rpfilter (1 references)
RETURN     all  --  anywhere             anywhere             rpfilter validmark

# table raw
iptables -t raw -L | grep rpfilter
```

as expected:

> make `conntrack` work: `rpfilter` must be in the `mangle` table, after the `conntrack` stage

</details>

RealityAnomaly reviewed (approved):
Looks good, would be nice to get this in.

milahu commented:
thanks for your review

i also wanted to show a notice to change config

```nix
{
networking.firewall.checkReversePath = false; # -> show notice
networking.firewall.checkReversePathNoticePR110197 = false; # -> hide notice
```

sample notice:

> please remove checkReversePath=false from your config. in most cases, this is a workaround for wireguard, which is not needed any longer. if you really want to disable rpfilter, set checkReversePathNoticePR110197=false; to hide this message

CitadelCore commented:
> thanks for your review
> 
> i also wanted to show a notice to change config
> 
> ```nix
> {
> networking.firewall.checkReversePath = false; # -> show notice
> networking.firewall.checkReversePathNoticePR110197 = false; # -> hide notice
> ```
> 
> sample notice:
> 
> > please remove checkReversePath=false from your config. in most cases, this is a workaround for wireguard, which is not needed any longer. if you really want to disable rpfilter, set checkReversePathNoticePR110197=false; to hide this message

Not sure this is a good idea. There are many legitimate reasons for disabling rpfilter, such as when running a router.

milahu commented:
rp_filter is "default on" for better security

its just a warning in the console, similar to [option x was renamed](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/rename.nix)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/844

zx2c4 commented:
Looks like somebody ran into this in the #wireguard IRC channel today. Any reason it's still pending?

ncfavier commented:
It's me, I'm somebody. I'll probably have a few comments on this, let me just test it first.

ncfavier reviewed (commented):
Tested on both client and server, works fine. Restarting `wg-quick` reinstalls the `--restore-mark` rule correctly in the first position since `-I` is used.

We should probably also remove

https://github.com/NixOS/nixpkgs/blob/d51fd786208f51fbaf36099778a942b915515adf/nixos/modules/services/networking/wg-quick.nix#L331-L333

since this PR fixes the issue.

ncfavier commented:
See https://github.com/firewalld/firewalld/issues/603 for a similar issue fixed by moving the rpfilter check to the mangle table.

milahu commented:
fixed 2 comments. should i rebase?

Commit:
nixos/firewall: move rpfilter from raw to mangle

fix wireguard (wg-quick)

netfilter packet flow:
raw.prerouting -> conntrack -> mangle.prerouting

rpfilter must be after conntrack
otherwise response packets are dropped

ncfavier commented:
> We should probably also remove
> 
> https://github.com/NixOS/nixpkgs/blob/d51fd786208f51fbaf36099778a942b915515adf/nixos/modules/services/networking/wg-quick.nix#L331-L333
> 
> since this PR fixes the issue.

I was suggesting doing this in this PR

ncfavier commented:
Last nit: please prefix the commit message with `nixos/firewall:`

ncfavier commented:
(and maybe succinctly explain why this is needed in the commit description)

milahu commented:
> See [firewalld/firewalld#603](https://github.com/firewalld/firewalld/issues/603) for a similar issue fixed by moving the rpfilter check to the mangle table.

also adds `--validmark` to rpfilter
but this is not needed to fix my use case (wireguard)

https://ipset.netfilter.org/iptables-extensions.man.html

> rpfilter
> 
> Performs a reverse path filter test on a packet. If a reply to the packet would be sent via the same interface that the packet arrived on, the packet will match. [...]
> 
> --validmark
> Also use the packets' nfmark value when performing the reverse path route lookup.

ncfavier commented:
We already [have](https://github.com/NixOS/nixpkgs/pull/110197/files#diff-9556b03f8bbc511b6349d6b68a4ee55aa2be8416c3ad3c4c08a5e94dcc594b23R120) `--validmark`

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/626

ncfavier commented:
@bobby285271 why remove the approvals label?

ncfavier commented:
Anyway, please merge 😅

milahu commented:
ping @SuperSandro2000 @NickCao @nrdxp 

guys, can we please merge this trivial patch?

SuperSandro2000 commented:
Sorry, I am not very confident with networking.

Artturin commented:
> I have a bot that runs https://github.com/nix-community/label-approved regularly and it removes approval labels [in some cases](https://github.com/nix-community/label-approved/blob/5842492714f254bd18101eabf6540918773cd4c7/src/label_approved/cli.py#L131), maybe we disable that behavior? /cc @Artturin

No, it was requested by @zowoq

> @Artturin I'm assuming the labels are from a bot? Having it label PRs as approved when the approval is outdated (the PR has new commits or force-pushed) is misleading.

https://github.com/NixOS/nixpkgs/pull/172542#issuecomment-1123697188
